superset-frontend/src/components/Datasource/DatasourceModal.tsx

```diff
@@ -96,39 +96,81 @@ const DatasourceModal: FunctionComponent<DatasourceModalProps> = ({
       currentDatasource.schema;
 
     setIsSaving(true);
-    SupersetClient.post({
-      endpoint: '/datasource/save/',
-      postPayload: {
-        data: {
-          ...currentDatasource,
-          cache_timeout:
-            currentDatasource.cache_timeout === ''
-              ? null
-              : currentDatasource.cache_timeout,
-          schema,
-          metrics: currentDatasource?.metrics?.map(
-            (metric: Record<string, unknown>) => ({
-              ...metric,
+    SupersetClient.put({
+      endpoint: `/api/v1/dataset/${currentDatasource.id}`,
+      jsonPayload: {
+        table_name: currentDatasource.table_name,
```

Comment:
Why are we not spreading as we were before? same with metric and column below.

Reply:
The old endpoint didn't have a rigid structure it accepted as a payload; it would just take a JSON string, parse it, and then set all keys that it was able to set on the target dataset, ignoring irrelevant keys.

The new endpoint is different; it accepts the payload defined by the `DatasetPutSchema`, which only includes the relevant keys that can be updated. If the payload contains any other keys, the endpoint 400's.  The datasource object here contains several of these other irrelevant keys, so now we have to extract only the ones that are needed.